src/potions.ts

```diff
@@ -417,3 +430,134 @@ export function bathroomFinance(embezzlers: number): void {
     }
   }
 }
+
+function triangleNumber(b: number, a = 0) {
+  return 0.5 * (b * (b + 1) - a * (a + 1));
+}
+
+class VariableMeatPotion {
+  potion: Item;
+  effect: Effect;
+  duration: number;
+  softcap: number; // Number of turns to cap out variable bonus
+  meatBonusPerTurn: number; // meat% bonus per turn
+  cappedMeatBonus: number;
+
+  constructor(
+    potion: Item,
+    softcap: number,
+    meatBonusPerTurn: number,
+    duration?: number,
+    effect?: Effect
+  ) {
+    this.potion = potion;
+    this.effect = effect ?? effectModifier(potion, "Effect");
+    this.duration = duration ?? numericModifier(potion, "Effect Duration");
+    this.softcap = softcap;
+    this.meatBonusPerTurn = meatBonusPerTurn;
+    this.cappedMeatBonus = softcap * meatBonusPerTurn;
+  }
+
+  use(quantity: number): boolean {
+    return use(quantity, this.potion);
+  }
+
+  price(historical: boolean): number {
+    // If asked for historical, and age < 14 days, use historical.
+    // If potion is not tradeable, use retrievePrice instead
+    return this.potion.tradeable
+      ? historical && historicalAge(this.potion) < 14
+        ? historicalPrice(this.potion)
+        : mallPrice(this.potion)
+      : retrievePrice(this.potion);
+  }
+
+  getOptimalNumberToUse(yachtzees: number, embezzlers: number): number {
+    const barfTurns = Math.max(0, estimatedTurns() - yachtzees - embezzlers);
+
+    const potionAmountsToConsider = [] as number[];
+    for (const fn of [Math.floor, Math.ceil]) {
+      for (const sc of [0, this.softcap]) {
+        for (const em of [0, embezzlers]) {
+          for (const bt of [0, barfTurns]) {
+            potionAmountsToConsider.push(fn((yachtzees + em + bt + sc) / this.duration));
+          }
+        }
+      }
+    }
+
+    const profits = [] as number[];
+    const profitsMap = new Map(
+      potionAmountsToConsider.map((n) => {
+        const profit = this.valueNPotions(n, yachtzees, embezzlers, barfTurns);
+        profits.push(profit);
```

Comment:
why not define profits as `Array.from(profitsMap.keys())`?

I would also probably not use a Map here, because we don't technically know that profits are unique."

So instead, maybe something like

```
const profitsFromPotions = potionAmountsToConsider.map((quantity) => ({ quantity, value: this.valueNPotions(quantity, yachtzees, embezzlers, barfTurns)"))
```

then later
```
const bestOption = profits.reduce((a, b) => a.value > b.value ? a : b);
if (bestOption.value > 0) {
  const potionsToUse = bestOption.quantity - Math.floor(haveEffect(this.effect) / this.duration);
  return Math.max(potionsToUse, 0);
}
```